### PR TITLE
[4.1][Feature] New field type - Toggle Field Visibility

### PR DESCRIPTION
**Pull request by OwenMelbz**
**Summary**

Hides/Shows other form groups based on a radio selection

**Preview**
backpack - toggle

**Usage**

``` php
//Controller->setUp()
$this->crud->addField([
    'label' => 'Is this a featured article?',
    'name' => 'featured', // can be a real db field, or unique name
    'type' => 'toggle',
    'options' => [ // same as radio, these act as the options, the key is the radio value
        0 => 'Not Featured',
        1 => 'Yes Featured'
    ],
    'hide_when' => [ // these fields hide (by name) when the key matches the radio value
        0 => ['featured_image', 'featured_title'],
        1 => ['basic_title']
    ],
    'default' => 0 // which option to select by default
]);
```

### CODE REVIEW

**Comment by OwenMelbz** [comment]
@tabacitu and maybe this :)

**Comment by OwenMelbz** [comment]
Hey @tabacitu - I've had a good few people recently asking about this feature, and as it's simple, wondering what the chances are of a merge for it soon? We're about to start another project and would be awesome to have a few more baked in :)

**Comment by MarcosBL** [comment]
Not sure how to propose this, don't know if I can make a PR over a PR, so I better comment it here, as it is a tiny change.

@OwenMelbz , iCheck doesn't propagate "change" events, so if you use it to style your forms, it will break toggle: it will hide fields on initial load correctly, but it will not work on the "change/click" events. 

In order to enable it for iCheck as well, we can listen for both native and iCheck events (**ifChecked**), and that solves it:

````javascript
$('input[data-field-toggle]').on('change ifChecked', function(){
````

image

image

**Comment by OwenMelbz** [comment]
Hmm, is a tricky one!

However, I would say we cannot cater for every plugin lol, otherwise we'd be never stopping!

My solution would be something along the lines of

- If you don't mind editing the view and adding in the event, thats fine, otherwise you should make your iCheck plugin or any other plugin fire the change event from their plugin.

I dont know how icheck works, but imagine this

```js
$('input[radio]').iCheck({
    onChange() {
        this.trigger('change');
    }
});
```

**Comment by MarcosBL** [comment]
You are of course right, I only commented because of AdminLTE being part of Backpack, and including the plugin (and because is a non breaking change, of course!), just in case someone found the same problem if they use your toggle template, so they can find the comment :)

**Comment by MarcosBL** [comment]
Just as aditional information, I managed to get it working without the event using this, pretty self-explanatory:

````javascript
if ( $.isFunction($.fn.iCheck) ) {
	$('input:not(.primary_list,.secondary_list)').on('ifChanged', function(event){
       		 $(event.target).trigger('change');
	}).iCheck({
		checkboxClass: 'icheckbox_flat-blue',
		radioClass: 'iradio_flat-blue',
	});
}
````

**Comment by jsiebach** [comment]
@OwenMelbz One thing that doesn't seem to work properly - a field that is toggled by multiple other fields.  Basically, field A is a toggle.  If a certain response is filled out, then show field B.  B is also a toggle.  B determines whether or not to show field C.

Field A is configured like:

    'hide_when' => [
      0 => ['B', 'C'],
      1 => []
    ]

Field B is configured like:

    'hide_when' => [
      0 => ['C'],
      1 => []
    ]

Then I go through this process:

1. Select "1" for "A"
  - B and C appear

2. Select "0" for "B"
  - C disappears

3. Toggle "A" to "0" and then back to "1"
  - C reappears, even though it should still be hidden due to it's dependency on B

**Comment by eyemiru** [comment]
Cool feature, waiting for it to be merged! :)

**Comment by jrpub** [comment]
Would be good to find a better way: Laravel is using VueJS for the client's side rendering (by default). The lack of convergence between Backpack and VueJS will lead to a problem one day or another, and it is already frustrating.
When I see the code proposed I don't think it is a good idea to merge it.

**Comment by lloy0076** [comment]
Integrating VueJS and Backpack would be a lot of work - I agree that Laravel *seems* to be going down the VueJS pathway but - and this is just my **own** opinion - I really don't trust Laravel to decide (next year, week, in Laravel 5.8) to choose some other shiny new JavaScript framework.

(I know I'm sounding old and grumpy :P).

**Comment by usmanhafeez147** [comment]
what is coding for toggle view?

**Comment by lloy0076** [comment]
@usmanhafeez147 - sorry I don't quite understand what you're asking in https://github.com/Laravel-Backpack/CRUD/pull/165#issuecomment-323520231?

File: src/resources/views/fields/toggle.blade.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
+<!-- radio toggle -->
+@php
+    $optionPointer = 0;
+    $optionValue = old($field['name']) ? old($field['name']) : (isset($field['value']) ? $field['value'] : (isset($field['default']) ? $field['default'] : '' ));
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: Ternaries in ternaries. Icck.

File: src/resources/views/fields/toggle.blade.php [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+<div @include('crud::inc.field_wrapper_attributes') >
+
+    <div>
+        <label>{!! $field['label'] !!}</label>
+    </div>
+
+    @if( isset($field['options']) && is_array($field['options']) )
+
+        @foreach ($field['options'] as $value => $label )
+            @php ($optionPointer++)
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: Is the `$loop` variable supported in the versions of Blade we support? https://laravel.com/docs/5.4/blade#loops

File: src/resources/views/fields/toggle.blade.php [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+            @if( isset($field['inline']) && $field['inline'] )
+
+            <label class="radio-inline" for="{{$field['name']}}_{{$optionPointer}}">
+                <input data-field-toggle="{{ json_encode($field['hide_when']) }}" type="radio" id="{{$field['name']}}_{{$optionPointer}}" name="{{$field['name']}}" value="{{$value}}" {{$optionValue!=='' && $optionValue == $value ? ' checked': ''}}> {!! $label !!}
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: `$optionValue !== ''` - notice the spaces (may as well be consistent with the following ==).

File: src/resources/views/fields/toggle.blade.php [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            </label>
+
+            @else
+
+            <div class="radio">
+                <label for="{{$field['name']}}_{{$optionPointer}}">
+                    <input data-field-toggle="{{ json_encode($field['hide_when']) }}" type="radio" id="{{$field['name']}}_{{$optionPointer}}" name="{{$field['name']}}" value="{{$value}}" {{$optionValue!=='' && $optionValue == $value ? ' checked': ''}}> {!! $label !!}
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: see above re spacing.

File: src/resources/views/fields/toggle.blade.php [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+</div>
+
+{{-- ########################################## --}}
+{{-- Extra CSS and JS for this particular field --}}
+{{-- If a field type is shown multiple times on a form, the CSS and JS will only be loaded once --}}
+@if ($crud->checkIfFieldIsFirstOfItsType($field, $fields))
+
+    {{-- FIELD CSS - will be loaded in the after_styles section --}}
+    @push('crud_fields_styles')
+
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: NL?

File: src/resources/views/fields/toggle.blade.php [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    @endpush
+
+    {{-- FIELD JS - will be loaded in the after_scripts section --}}
+    @push('crud_fields_scripts')
+    <script>
+        jQuery(document).ready(function($){
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: Is there a reason we're spelling jQuery here but using `$` in the function?

File: src/resources/views/fields/toggle.blade.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+
+            window.$hiddenFields = window.$hiddenFields || {};
+
+            var $toggle = function( $radio ){
+
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: NL?

File: src/resources/views/fields/toggle.blade.php [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                $hideWhen = $radio.data('field-toggle'),
+                $value    = $radio.val(),
+                $radioSet = $radio.attr('name');
+
+                $hiddenFields[ $radioSet ] = $hiddenFields[ $radioSet ] || [];
+
+                if( Object.keys($hiddenFields[ $radioSet ]).length ){
+                    $.each($hiddenFields[ $radioSet ], function(idx, field){
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: Ok, I have to ask, everything has gotten the evil $ that isn't needed in JS, why not the function parameters? :imp:

File: src/resources/views/fields/toggle.blade.php [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+                        field.show();
+                    });
+                    $hiddenFields[ $radioSet ] = [];
+                }
+
+                if( typeof $hideWhen[ $value ] !== undefined ){
+                    $.each($hideWhen[ $value ], function(idx, name){
```
Comment (lloy0076) [inline src/resources/views/fields/toggle.blade.php]: See my impish questin above :P

**Comment by OwenMelbz** [comment]
Ha just tried to use this, forgetting it didn't get merged.

@tabacitu what was your opinion on all this, would it be something you could add soon?

**Comment by tabacitu** [comment]
Sorry @OwenMelbz - yes, I think you're right, I think it's about time we implement a solution for this problem. I'll test & review this PR shortly, along with https://github.com/Laravel-Backpack/CRUD/pull/1257 and see which one to merge. I really need to test this carefully, and see if interacts ok with tabs & everything.

**Comment by Tarpsvo** [comment]
It's been 84 years.

**Comment by OwenMelbz** [comment]
@Tarpsvo I believe you missed out an extra 1, 184 years :D

**Comment by Tarpsvo** [comment]
I copied the field to my project and it works like a charm - thank you for this! 👍

**Comment by srigurubyo** [comment]
I believe this is for check (toggle) kind of fields.  

I have a scenario where a file should be shown as dropdown or text based on the value of another field.

Example: User to select states (California, North Carolina, etc) if the selected country is USA.  If any other country is selected (other than USA), user should be shown only a editable text field (instead of dropdown).

Please let me know if this feature can help me achieve this.

**Comment by tabacitu** [comment]
Wow this is probably one of the oldest PRs of yours that I haven't merged @OwenMelbz . 2016 - man, time passes fast.

I've decided this doesn't make the cut as a core feature, but I'd love to have it as an add-on. So I've opened https://github.com/Laravel-Backpack/addons/issues/11 in the hopes that maybe someone else likes this idea and wants to create an add-on and maintain it. If you @OwenMelbz don't want to, maybe someone else who's using it will. If so, let's move the conversation there, in https://github.com/Laravel-Backpack/addons/issues/11 .

Cheers!

**Comment by sandeepchowdary7** [comment]
Hello,

Is this PR merged? is the `toggle` type available in backpack 4.1 V??

**Comment by tabacitu** [comment]
@sandeepchowdary7 no - the feature is still in development. I've just pushed a PR with the most likely solution yesterday - you can see it here - https://github.com/Laravel-Backpack/CRUD/pull/4312

**Pull request closed without merging**